opm/material/fluidsystems/BlackOilFluidSystem.hpp

```diff
@@ -1071,42 +1077,192 @@ class BlackOilFluidSystem : public BaseFluidSystem<Scalar, BlackOilFluidSystem<S
+
+                return waterEnergy;
+            }
+            is_mixing = true;
+
+        }
+        case gasPhaseIdx: {
+            if(!gasPvt_->mixingEnergy()){
+                const auto gasEnergy =
+                    gasPvt_->internalEnergy(regionIdx, T, p,
+                                            BlackOil::template getRv_<ThisType, FluidState, LhsEval>(fluidState, regionIdx),
+                                            BlackOil::template getRvw_<ThisType, FluidState, LhsEval>(fluidState, regionIdx));
+
+                return gasEnergy;
+            }
+            is_mixing = true;
+        }
+        }
+        assert(is_mixing==true);
+        const auto& energy = internalMixingTotalEnergy<FluidState,LhsEval>(fluidState, phaseIdx, regionIdx);
+        const auto& phase_density = density<FluidState,LhsEval>(fluidState, phaseIdx, regionIdx);
+        return energy/phase_density;
+    }
+
+
+    template <class FluidState, class LhsEval = typename FluidState::Scalar>
+    static LhsEval internalMixingTotalEnergy(const FluidState& fluidState,
+                                             unsigned phaseIdx,
+                                             unsigned regionIdx)
     {
         assert(phaseIdx <= numPhases);
         assert(regionIdx <= numRegions());
+        const LhsEval& p = decay<LhsEval>(fluidState.pressure(phaseIdx));
+        const LhsEval& T = decay<LhsEval>(fluidState.temperature(phaseIdx));
+        const LhsEval& saltConcentration = BlackOil::template getSaltConcentration_<ThisType, FluidState, LhsEval>(fluidState, regionIdx);
+        // to avoid putting all thermal into the interface of the multiplexer
+        switch (phaseIdx) {
+        case oilPhaseIdx: {
+            auto oilEnergy = oilPvt_->internalEnergy(regionIdx, T, p,
+                                                     BlackOil::template getRs_<ThisType, FluidState, LhsEval>(fluidState, regionIdx));
+            assert(oilPvt_->mixingEnergy());
+            //mixing energy adsed
+            if (enableDissolvedGas()) {
+                // miscible oil
+                const LhsEval& Rs = BlackOil::template getRs_<ThisType, FluidState, LhsEval>(fluidState, regionIdx);
+                const LhsEval& bo = oilPvt_->inverseFormationVolumeFactor(regionIdx, T, p, Rs);
+                const auto& gasEnergy =
+                    gasPvt_->internalEnergy(regionIdx, T, p,
+                                            BlackOil::template getRv_<ThisType, FluidState, LhsEval>(fluidState, regionIdx),
+                                            BlackOil::template getRvw_<ThisType, FluidState, LhsEval>(fluidState, regionIdx));
+                const auto hVapG = gasPvt_->hVap(regionIdx);// pressure correction ? assume equal to energy change
+                return
+                    oilEnergy*bo*referenceDensity(oilPhaseIdx, regionIdx)
+                    + (gasEnergy-hVapG)*Rs*bo*referenceDensity(gasPhaseIdx, regionIdx);
+            }
```

Comment:
Why is hVapG minus? Later it is pluss? I guess this should be consistent.

Reply:
I will change this to + and the other to minus. I.e. energy of gas is given in the gasphase while energy of liquid is given in the liquid phase. Does this make sence?